v2/http-middleware/ratelimit/ratelimit.go

```diff
@@ -0,0 +1,115 @@
+package ratelimit
+
+import (
+	"fmt"
+	"github.com/SKF/go-utility/v2/http-middleware/util"
+	"github.com/gomodule/redigo/redis"
+	"github.com/gorilla/mux"
+	"net/http"
+	"time"
+)
+
+type Store interface {
+	Incr(string) (int, error)
+}
+
+type EndpointConfig struct {
+	path   Request
+	config []Config
+}
+
+type Config struct {
+	requestPerMinute int
+	getKeyFunc       func(req *http.Request) (string, error)
+}
+
+type Request struct {
+	method string
+	path   string
+}
+
+type Limiter struct {
+	store   Store
+	configs map[Request][]Config
+}
+
+func SetStore(s Store) Limiter {
+	return Limiter{
+		store:   s,
+		configs: map[Request][]Config{},
+	}
+}
+
+func (s *Limiter) Configure(config EndpointConfig) {
+	s.configs[config.path] = config.config
+}
+
+type redisStore struct {
+	url string
+}
+
+func (s *redisStore) Incr(key string) (int, error) {
+	conn, err := redis.Dial("tcp", s.url)
+	if err != nil {
+		return -1, err
+	}
+	defer conn.Close()
+
+	cnt, err := redis.Int(conn.Do("INCR", key))
+	if err != nil {
+		return -1, err
+	}
+
+	_, err = conn.Do("EXPIRE", key, 59)
+	if err != nil {
+		return -1, err
+	}
+
+	return cnt, nil
+}
+
+func GetRedisStore(url string) Store {
+	return &redisStore{url: "localhost:6379"}
+}
+
+func (s *Limiter) Middleware() mux.MiddlewareFunc {
+	if s.store == nil {
+		panic("store is not configured")
+	}
+
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			_, span := util.StartSpanNoRoot(req.Context(), "RateLimitMiddleware/Handler")
+			defer span.End()
+
+			now := time.Now()
+
+			cfgs, ok := s.configs[Request{method: req.Method, path: req.URL.Path}]
+			if ok {
+				for _, config := range cfgs {
+					key, err := config.getKeyFunc(req)
+					key = fmt.Sprintf("%s:%d", key, now.Minute())
```

Comment:
Clever trick using the current minute as part of the key! Since it is not really a rolling time window, it gives a misbehaving user the possibility to do 2*rate limit requests in a rolling one minute window if the attack occurs around a minute change, but I guess that's not really a problem since after that initial burst of requests the rate limit would work as expected.

A problem however, is that since the key is made up of path, method and minute, this rate limiting will enable an attacker to shut down an endpoint for all users by constantly exceeding the rate limit. I assume that what you want is to just lock out the misbehaving user, in which case the source of the request somehow needs to be part of the key; maybe the `RemoteAddr` field of the `http.Request` could be used?

Reply:
Since we providing one ore several custom getKey functions for each "Request" We can include whatever we like in the key.

In the sign in for example We will include the email.
In the other more regular endpoints we can use for example uuid or client-ID + path.

To only use the path was just a simplification in the test cases.

Reply:
Ah, now I see; looks nice!